### PR TITLE
Feat: Upgrade LangChain to Version 0.3

### PR DESCRIPTION
**Pull request by Pouyanpi**
This pull request makes the necessary changes to upgrade LangChain to version 0.3. 

1. **Update Pydantic Import Paths**:
   - Update import paths for `pydantic.v1` to ensure compatibility with the latest version of Pydantic.
   - Suppress specific warnings related to protected namespaces in Pydantic models by adding filter to ignore them. 

2. **Update LangChain Version Constraints**:
   - Update version constraints for `langchain`, `langchain-core`, and `langchain-community` to allow versions up to <0.4.0.

3. **Refactor Import Paths for BaseLLM**:
   - Update import paths for `BaseLLM` to `langchain_core.language_models.llms` across multiple files to align with the latest version of LangChain.
   
- [x] automated test passed
- [x] run e2e tests using various configs

### CODE REVIEW

**Comment by Pouyanpi** [comment]
It resolves #782 and #767 .

**Review by cparisien — APPROVED** [review]

File: nemoguardrails/llm/providers/providers.py [hunk 1]
```diff
@@ -23,30 +23,37 @@
...
+    category=UserWarning,
+    module=r"pydantic\._internal\._fields",
+)
 log = logging.getLogger(__name__)
 
-# Initialize the providers with the default ones, for now only NeMo LLM.
+# Initialize the providers with the default ones
 # We set nvidia_ai_endpoints provider to None because it's only supported if `langchain_nvidia_ai_endpoints` is installed.
-_providers: Dict[str, Type[BaseLanguageModel]] = {
+_providers: Dict[str, Optional[Type[BaseLLM]]] = {
```
Comment (trebedea) [inline nemoguardrails/llm/providers/providers.py:56]: Do we need here the Optional?

**Review by trebedea — APPROVED** [review]
LGTM 👍

**Comment by aortu22** [comment]
> When is this develop going to be release? I believe it is kind of urgent

Maybe a v 0.10.2 can be released as a patch to cover this issue

**Pull request merged**